jmh-benchmarks/src/main/java/org/apache/kafka/jmh/timeline/TimelineHashMapBenchmark.java

```diff
@@ -0,0 +1,87 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.jmh.timeline;
+
+import org.apache.kafka.common.utils.LogContext;
+import org.apache.kafka.timeline.SnapshotRegistry;
+import org.apache.kafka.timeline.TimelineHashMap;
+import org.openjdk.jmh.annotations.Benchmark;
+import org.openjdk.jmh.annotations.BenchmarkMode;
+import org.openjdk.jmh.annotations.Fork;
+import org.openjdk.jmh.annotations.Measurement;
+import org.openjdk.jmh.annotations.Mode;
+import org.openjdk.jmh.annotations.OutputTimeUnit;
+import org.openjdk.jmh.annotations.Scope;
+import org.openjdk.jmh.annotations.State;
+import org.openjdk.jmh.annotations.Warmup;
+
+import java.util.HashMap;
+import java.util.concurrent.TimeUnit;
+
+@State(Scope.Benchmark)
+@Fork(value = 1)
+@Warmup(iterations = 3)
+@Measurement(iterations = 10)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.MILLISECONDS)
+
+public class TimelineHashMapBenchmark {
+    private final static int NUM_ENTRIES = 1_000_000;
+
+    @Benchmark
+    public void testAddEntriesInHashMap() {
+        SnapshotRegistry snapshotRegistry = new SnapshotRegistry(new LogContext());
+        HashMap<Integer, String> map = new HashMap<>(NUM_ENTRIES);
+        for (int i = 0; i < NUM_ENTRIES; i++) {
+            int key = (int)(0xffffffff & ((i * 2862933555777941757L) + 3037000493L));
+            map.put(key, String.valueOf(key));
+        }
+    }
+
+    @Benchmark
+    public void testAddEntriesInTimelineMap() {
+        SnapshotRegistry snapshotRegistry = new SnapshotRegistry(new LogContext());
+        TimelineHashMap<Integer, String> map =
+            new TimelineHashMap<>(snapshotRegistry, NUM_ENTRIES);
+        for (int i = 0; i < NUM_ENTRIES; i++) {
+            int key = (int)(0xffffffff & ((i * 2862933555777941757L) + 3037000493L));
+            map.put(key, String.valueOf(key));
+        }
+    }
+
+    @Benchmark
+    public void testAddEntriesWithSnapshots() {
+        SnapshotRegistry snapshotRegistry = new SnapshotRegistry(new LogContext());
+        TimelineHashMap<Integer, String> map =
+            new TimelineHashMap<>(snapshotRegistry, NUM_ENTRIES);
+        long epoch = 0;
+        int j = 0;
+        for (int i = 0; i < NUM_ENTRIES; i++) {
+            int key = (int)(0xffffffff & ((i * 2862933555777941757L) + 3037000493L));
+            if (j > 10 && key % 3 == 0) {
```

Comment:
Not sure what the intention of of the logic here. It seems that we skip the snapshotting for slightly less than 10% of the time?

Reply:
the intention is to have each snapshot have around 10-13 entries on average

Reply:
Hmm, does the code achieve this? It seem that at least 2/3 of the iterations will trigger the creation of a new snapshot?

Reply:
oh, sorry, I put the snapshot creation / deleting in the wrong branch of the "if"